lib/fluent/plugin/out_azure-storage-append-blob.rb

```diff
@@ -3,122 +3,157 @@
-      config_param :azure_container, :string, :default => nil
-      config_param :azure_object_key_format, :string, :default => "%{path}%{time_slice}-%{index}.log"
-      config_param :auto_create_container, :bool, :default => true
-      config_param :format, :string, :default => DEFAULT_FORMAT_TYPE
-      config_param :time_slice_format, :string, :default => '%Y%m%d'
-  
+
+      config_param :path, :string, default: ''
+      config_param :azure_storage_account, :string, default: nil
+      config_param :azure_storage_access_key, :string, default: nil, secret: true
+      config_param :azure_storage_sas_token, :string, default: nil, secret: true
+      config_param :azure_container, :string, default: nil
+      config_param :use_msi, :bool, default: false
+      config_param :azure_object_key_format, :string, default: '%{path}%{time_slice}-%{index}.log'
+      config_param :auto_create_container, :bool, default: true
+      config_param :format, :string, default: DEFAULT_FORMAT_TYPE
+      config_param :time_slice_format, :string, default: '%Y%m%d'
+      config_param :localtime, :bool, default: false
+
       config_section :format do
         config_set_default :@type, DEFAULT_FORMAT_TYPE
       end
-  
+
       config_section :buffer do
         config_set_default :chunk_keys, ['time']
         config_set_default :timekey, (60 * 60 * 24)
       end
-  
+
       attr_reader :bs
-  
+
       def configure(conf)
         super
-  
+
         @formatter = formatter_create
-  
-        if @localtime
-          @path_slicer = Proc.new {|path|
-            Time.now.strftime(path)
-          }
-        else
-          @path_slicer = Proc.new {|path|
-            Time.now.utc.strftime(path)
-          }
-        end
-  
-        if @azure_container.nil?
-          raise ConfigError, 'azure_container is needed'
-        end
+
+        @path_slicer = if @localtime
+                         proc do |path|
+                           Time.now.strftime(path)
+                         end
+                       else
+                         proc do |path|
+                           Time.now.utc.strftime(path)
+                         end
+                       end
+
+        raise ConfigError, 'azure_storage_account needs to be specified' if @azure_storage_account.nil?
+
+        raise ConfigError, 'azure_container needs to be specified' if @azure_container.nil?
 
         if @azure_storage_access_key.nil? && @azure_storage_sas_token.nil?
-          raise ConfigError, "either 'azure_storage_access_key' or 'azure_storage_sas_token' parameter must be provided"
+          log.info 'Using MSI since neither azure_storage_access_key nor azure_storage_sas_token was provided.'
+          @use_msi = true
         end
       end
-  
+
       def multi_workers_ready?
         true
       end
-  
+
+      def get_access_token
+        access_key_request = Faraday.new('http://169.254.169.254/metadata/identity/oauth2/token?' \
```

Comment:
this looks good for msi support, but i think the api version should be configurable

Reply:
Sure, I'll add that.